### ISSUE TITLE
Cannot replicate demo 8: No session_state in extended callback

### ISSUE THREAD

**Issue by andrealorenzon** [GibbsConsulting/django-plotly-dash#356]
Problem:
I have no success in replicating demo-8. In detail, I cannot access to session_state or other kwarg mentioned [here](https://django-plotly-dash.readthedocs.io/en/latest/extended_callbacks.html).
All other callback feature works, like update of other Graphs and Divs.

My files:

## urls.py

```python
from django.urls import path
from . import views

from viewer.dash_apps.finished_apps import djangosessionstate

urlpatterns = [
    path(
        "djangosessionstate",
        views.djangosessionstate,
        {"template_name": "viewer/djangosessionstate.html"},
        name="cbt",
    ),
```

## views.py

```python

def djangosessionstate(request, template_name, **kwargs):

    session = request.session

    demo_count = session.get('django_plotly_dash', {})

    ind_use = demo_count.get('ind_use', 0)
    ind_use += 1
    demo_count['ind_use'] = ind_use

    context = {'ind_use' : ind_use}

    session['django_plotly_dash'] = demo_count

    return render(request, template_name=template_name, context=context)

```

## djangosessionstate.py

```python
from dash import html
from dash import dcc
from dash.dependencies import Input, Output
from django_plotly_dash import DjangoDash
import dash_bootstrap_components as dbc
import dash

external_stylesheet = [
    {
        "href": "https://codepen.io/chridddyp/pen/bWLwgP.css",
        "rel": "stylesheet",
        "crossorigin": "anonymous",
    }
]

dis = DjangoDash("DjangoSessionState", add_bootstrap_links=True)

dis.layout = html.Div(
    [
        dbc.Alert("This is an alert", id="base-alert", color="primary"),
        dbc.Alert(children="Danger", id="danger-alert", color="danger"),
        dbc.Button("Update session state", id="update-button", color="warning"),
        ]
    )

@dis.expanded_callback(
    dash.dependencies.Output("base-alert", 'children'),
    [dash.dependencies.Input('danger-alert', 'children'),]
    )
def session_demo_danger_callback(da_children, session_state=None, **kwargs):
    'Update output based just on state'
    if not session_state:
        return "Session state not yet available"

    return "Session state contains: " + str(session_state.get('bootstrap_demo_state', "NOTHING")) + " and the page render count is " + str(session_state.get("ind_use", "NOT SET"))

#pylint: ignore=unused-argument
@dis.expanded_callback(
    dash.dependencies.Output("danger-alert", 'children'),
    [dash.dependencies.Input('update-button', 'n_clicks'),]
    )
def session_demo_alert_callback(n_clicks, session_state=None, **kwargs):
    'Output text based on both app state and session state'
    if session_state is None:
        raise NotImplementedError("Cannot handle a missing session state")
    csf = session_state.get('bootstrap_demo_state', None)
    if not csf:
        csf = dict(clicks=0, overall=0)
    else:
        csf['clicks'] = n_clicks
        if n_clicks is not None and n_clicks > csf.get('overall_max',0):
            csf['overall_max'] = n_clicks
    session_state['bootstrap_demo_state'] = csf
    return "Button has been clicked %s times since the page was rendered" %n_clicks
```

## djangosessionstate.html

```html
{% extends 'base.html' %}
{% load static %}

{% block content %}
{% load plotly_dash %}

<div><h1>CallbackTest</h1></div>


<div class="{% plotly_class name='DjangoSessionState'%} card" style="height: 100%; width: 100%">
    {% csrf_token %}
    {% plotly_app name='DjangoSessionState' ratio=1 %}
</div>

{% endblock %}
```

base.html correctly loads plotly.js, jquery, and css's and works with djangodash apps that doesn't use session.

The error in browser's console is:

![image](https://user-images.githubusercontent.com/43496697/134661788-32358b8b-9d8e-4f20-835c-1a417f086b8b.png)

Temporary workaround: I declared a singleton object, stuck session data inside it, and read it from inside the app.

**Comment by GibbsConsulting** [comment]
Did you specify all of the correct settings in `settings.py`? Are you able to share the content of `kwargs` in the callback where you raise your error?

**Comment by andrealorenzon** [comment]
Kwargs is empty inside the callback.

My settings.py is:

```python
"""
Django settings for h5nuvola project.

Generated by 'django-admin startproject' using Django 3.2.7.

For more information on this file, see
https://docs.djangoproject.com/en/3.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.2/ref/settings/
"""

from pathlib import Path

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = 'django-insecure-74a+h^o^vs9s7yxs2^b!9%57*!=f7t2yrg^zl8m0i1$^yf+(75'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = [
    'viewer',
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django_plotly_dash.apps.DjangoPlotlyDashConfig',
    'channels',  # redis
    'channels_redis',

]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'h5nuvola.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            BASE_DIR / "templates"
        ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'h5nuvola.wsgi.application'

# TODO asgi application server cfg
ASGI_APPLICATION = 'h5nuvola.asgi.application'


# Backup local Database
# DATABASES = {
#     'default': {
#         'ENGINE': 'django.db.backends.sqlite3',
#         'NAME': BASE_DIR / 'db.sqlite3',
#     }
# }
# TODO database settings, makemigrations
# SECRET
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': 'h5nuvola',
        'USER': 'h5nuvola',
        'PASSWORD': 'h5nuvola',
        'HOST': 'localhost'
    }
}

# Password validation
# https://docs.djangoproject.com/en/3.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'CET'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.2/howto/static-files/

# TODO collectstatics
STATIC_URL = '/static/'
STATIC_ROOT = "staticfiles"
STATICFILES_LOCATION = "static"
STATICFILES_DIRS = [
    BASE_DIR / 'viewer' / 'static',
    BASE_DIR / 'h5nuvola' / 'static'

]
# Default primary key field type
# https://docs.djangoproject.com/en/3.2/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'


# DPD - configs

X_FRAME_OPTIONS = 'SAMEORIGIN'

# TODO start Redis server:
# docker pull redis:4
# docker run -p 6379:6379 -d redis

# TODO bootstrap template
CRISPY_TEMPLATE_PATH = 'bootstrap4'

# TODO channel
CHANNEL_LAYERS = {
    'default': {
        'BACKEND': 'channels_redis.core.RedisChannelLayer',
        'CONFIG':  {
            'hosts': [('127.0.0.1', 6379), ],  # TODO redis server
        }
    }
}

STATICFILES_FINDERS = [
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    'django_plotly_dash.finders.DashAssetFinder',
    'django_plotly_dash.finders.DashComponentFinder'
]

PLOTLY_COMPONENTS = [
    'dash_core_components',
    'dash_html_components',
    'dash_renderer',

    'dpd_components',
]
# TODO set cookie expiration age
SESSION_COOKIE_AGE = 60*60*24*30*6

# TODO # SECRET edit in deployment
H5NUVOLA_BASEDIR = BASE_DIR / 'h5nuvola'
```

**Comment by andrealorenzon** [comment]
Additional info:
# pip freeze
```bash
aioredis==1.3.1
asgiref==3.4.1
async-timeout==3.0.1
attrs==21.2.0
autobahn==21.2.1
Automat==20.2.0
beautifulsoup4==4.10.0
black==21.9b0
bootstrap4==0.1.0
Brotli==1.0.9
cffi==1.14.6
channels==3.0.4
channels-redis==3.3.0
click==8.0.1
constantly==15.1.0
cryptography==3.4.8
daphne==3.0.2
dash==2.0.0
dash-bootstrap-components==0.13.1
dash-core-components==2.0.0
dash-html-components==2.0.0
dash-renderer==1.9.1
dash-table==5.0.0
dataclasses==0.8
Django==3.2.7
django-bootstrap4==3.0.1
django-plotly-dash==1.6.5
django-redis==5.0.0
dpd-components==0.1.0
dpd-static-support==0.0.5
flake8==3.9.2
Flask==2.0.1
Flask-Compress==1.10.1
hiredis==2.0.0
hyperlink==21.0.0
idna==3.2
importlib-metadata==4.8.1
incremental==21.3.0
itsdangerous==2.0.1
Jinja2==3.0.1
MarkupSafe==2.0.1
mccabe==0.6.1
msgpack==1.0.2
mypy-extensions==0.4.3
pathspec==0.9.0
pkg-resources==0.0.0
platformdirs==2.4.0
plotly==5.3.1
psycopg2==2.9.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycodestyle==2.7.0
pycparser==2.20
pyflakes==2.3.1
pyOpenSSL==20.0.1
pytz==2021.1
redis==3.5.3
regex==2021.9.24
semantic-version==2.8.5
service-identity==21.1.0
setuptools-rust==0.12.1
six==1.16.0
soupsieve==2.2.1
sqlparse==0.4.2
tenacity==8.0.1
toml==0.10.2
tomli==1.2.1
Twisted==21.7.0
txaio==21.2.1
typed-ast==1.4.3
typing-extensions==3.10.0.2
Werkzeug==2.0.1
zipp==3.5.0
zope.interface==5.4.0
```

**Comment by GibbsConsulting** [comment]
Can you try with a version of Dash < 2.0?

I suspect that this (ie not yet compatible with newer versions of Dash) could be the problem (there is #354 also)

It looks like the installed package does not restrict the version of Dash installed so #357 has been raised

**Comment by andrealorenzon** [comment]
```
Successfully installed dash-1.21.0 dash-core-components-1.17.1 dash-html-components-1.1.4 dash-table-4.12.0 future-0.18.2
```
Now it seems to work. I'll check if everything is ok then mark it solved. Thank you for your support.

![image](https://user-images.githubusercontent.com/43496697/135042197-d39f7c90-7585-4d87-9648-cb8b586a79c6.png)

**Comment by andrealorenzon** [comment]
It doesn't depend only from Dash version, apparently. I'm having the same issue on another machine with Dash v.1.21

**Comment by GibbsConsulting** [comment]
Could this be related to #349 ? Are you able to try with a dash version < 1.20

**Comment by andrealorenzon** [comment]
No, it's not related, I think.

The same code, with the same virtual environment, works on a machine, while on another one gives the following error:

```
HTTP POST /django_plotly_dash/app/SimpleExample/_dash-update-component 500 [0.02, 127.0.0.1:47864]
HTTP GET /viewer/?file=/home/al/hdf5_test_files/test3d2.h5 200 [0.04, 127.0.0.1:47864]
HTTP GET /django_plotly_dash/app/SimpleExample/ 200 [0.01, 127.0.0.1:47864]
HTTP GET /django_plotly_dash/app/SimpleExample/_dash-layout 200 [0.02, 127.0.0.1:47864]
HTTP GET /django_plotly_dash/app/SimpleExample/_dash-dependencies 200 [0.02, 127.0.0.1:47862]
Internal Server Error: /django_plotly_dash/app/SimpleExample/_dash-update-component
Traceback (most recent call last):
  File "/home/al/.local/lib/python3.9/site-packages/asgiref/sync.py", line 482, in thread_handler
    raise exc_info[1]
  File "/home/al/.local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 38, in inner
    response = await get_response(request)
  File "/home/al/.local/lib/python3.9/site-packages/django/core/handlers/base.py", line 233, in _get_response_async
    response = await wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/al/.local/lib/python3.9/site-packages/asgiref/sync.py", line 444, in __call__
    ret = await asyncio.wait_for(future, timeout=None)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
    return await fut
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/al/.local/lib/python3.9/site-packages/asgiref/sync.py", line 486, in thread_handler
    return func(*args, **kwargs)
  File "/home/al/.local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/django_plotly_dash/views.py", line 74, in update
    return _update(request, ident, stateless, **kwargs)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/django_plotly_dash/views.py", line 91, in _update
    resp = app.dispatch_with_args(request_body, arg_map)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/django_plotly_dash/dash_wrapper.py", line 700, in dispatch_with_args
    res = callback(*args, **argMap)
  File "/home/linuxbrew/.linuxbrew/opt/python@3.9/lib/python3.9/site-packages/dash/_callback.py", line 151, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
  File "/home/al/apps/h5nuvola/viewer/dash_apps/finished_apps/simpleexample.py", line 46, in callback_c
    selected_file = kwargs["request"].session.get("opened_file", None)
KeyError: 'request'
HTTP POST /django_plotly_dash/app/SimpleExample/_dash-update-component 500 [0.02, 127.0.0.1:47864]
```

The differences between the two machines could be in the postgres server. Where it doesn't works, the database runs in a local docker container, while where it does it is installed. Separate testing proves that both db work properly. Both runs redis as a docker container.

Where it doesn't work:

```
al@thanos:~$ sudo docker ps
[sudo] password for al:         
CONTAINER ID   IMAGE      COMMAND                  CREATED      STATUS      PORTS                                       NAMES
0a9c619ff16c   postgres   "docker-entrypoint.s…"   8 days ago   Up 8 days   0.0.0.0:5432->5432/tcp, :::5432->5432/tcp   dev-postgres
2b6867afdb25   redis      "docker-entrypoint.s…"   8 days ago   Up 8 days   0.0.0.0:6379->6379/tcp, :::6379->6379/tcp   tender_driscoll
```

**Comment by andrealorenzon** [comment]
**Closed because duplicate of #349**
